src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationTokenAcquisition.cs

```diff
@@ -87,16 +106,29 @@ private async Task<IConfidentialClientApplication> GetOrCreateApplication()
         }
 
         /// <inheritdoc/>
-        public async Task<string> GetAccessTokenForUserAsync(
+        public Task<string> GetAccessTokenForUserAsync(
             IEnumerable<string> scopes,
             string? tenantId = null,
             string? userFlow = null,
             ClaimsPrincipal? user = null,
             TokenAcquisitionOptions? tokenAcquisitionOptions = null)
         {
-            string accessToken = GetAccessToken(CurrentHttpContext?.Request.Headers);
+            var httpContext = CurrentHttpContext;
+            string accessToken;
+            if (httpContext != null)
+            {
+                // Need to lock due to https://docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?#do-not-access-httpcontext-from-multiple-threads
+                lock (httpContext)
```

Comment:
Is this the pattern that should be used throughout the entire library when dealing with HttpContext?

From what I see, if 2 different requests occur at the same time, the HttpContext is different and no locking occurs.

If the same top-level request ends up accessing the Headers twice in this method, then the locking occurs. Is this scenario even possible?

Reply:
According to the linked docs page, it is something we should protect against. The http context is completely not thread safe and neither are the request/response properties of it. Even if this library does not access the httpContext in parallel, external libraries can. I took a look at the HttpContext source code and found that even reading certain fields in parallel can corrupt it due to how it lazy adds HttpFeatures to an internal collection.